src/consumer/subscriptionState.js

```diff
@@ -1,26 +1,73 @@
+const { KafkaJSNonRetriableError } = require('../errors')
+
 module.exports = class SubscriptionState {
   constructor() {
-    this.pausedTopics = new Set()
+    this.pausedPartitionsByTopic = {}
   }
 
   /**
-   * @param {Array<string>} topics
+   * @param {Array<TopicPartitions>} topicPartitions Example: [{ topic: 'topic-name', partitions: [1, 2] }]
    */
-  pause(topics = []) {
-    topics.forEach(topic => this.pausedTopics.add(topic))
+  pause(topicPartitions = []) {
+    topicPartitions.forEach(({ topic, partitions }) => {
+      const pausedForTopic = this.pausedPartitionsByTopic[topic] || {
+        topic,
+        partitions: new Set(),
+        all: false,
+      }
+
+      if (typeof partitions === 'undefined') {
+        pausedForTopic.partitions.clear()
+        pausedForTopic.all = true
+      } else if (Array.isArray(partitions)) {
+        partitions.forEach(partition => pausedForTopic.partitions.add(partition))
+        pausedForTopic.all = false
+      }
+
+      this.pausedPartitionsByTopic[topic] = pausedForTopic
+    })
   }
 
   /**
-   * @param {Array<string>} topics
+   * @param {Array<TopicPartitions>} topicPartitions Example: [{ topic: 'topic-name', partitions: [1, 2] }]
    */
-  resume(topics = []) {
-    topics.forEach(topic => this.pausedTopics.delete(topic))
+  resume(topicPartitions = []) {
+    topicPartitions.forEach(({ topic, partitions }) => {
+      const pausedForTopic = this.pausedPartitionsByTopic[topic] || { topic, partitions: new Set() }
+
+      if (typeof partitions === 'undefined') {
+        pausedForTopic.partitions.clear()
+        pausedForTopic.all = false
+      } else if (Array.isArray(partitions) && !pausedForTopic.all) {
+        partitions.forEach(partition => pausedForTopic.partitions.delete(partition))
+      } else if (Array.isArray(partitions) && pausedForTopic.all) {
+        // TODO: consider whether we should actively track active topics, rather than paused ones, as to avoid this,
```

Comment:
We'll probably want to either do something about this, or document it elsewhere. I don't think it's a proven use case yet, so throwing for now is fine.

Reply:
I feel like we should probably make a note of this in an issue or something and bundle it into some future API breaking change instead, because the current behavior is pretty confusing (but totally the right call, given backwards compatibility).

Reply:
Do you mean the API difference between calling with or withouts `partitions` specified? This comment was about the error we have to throw when we've previously paused all topics by not specifying `partitions`, to then try to resume a single partition. My idea so far has been that people use either full topic pausing, or per partition, but if this thrown Error is too confusing, we can go with bringing the current member assignment into `SubscriptionState`, which would allow mixing the two. 

> This could be solved by not only keeping a collection of paused topics, but also resumed. An alternative would be for the SubscriptionState to own the state describing what topics and partitions are assigned (currently in consumerGroup.topics and consumerGroup.memberAssignment), which given it's name suggests that might be the right thing to do anyway. That approach would also clean up consumerGroup.fetch which now has to join the paused with the assigned manually (although it's the only place where that needs to happen, so perhaps that's fine).

Maybe that's just the way to go anyway, as it would probably clean things up a bit, too. I'll have a look at how much is involved by having a small go at it.

Reply:
Yeah, I was talking about the general problem of calling the public API with or without partitions. I'm not so sure that the assumption that people will either pause entire topics or individual partitions is valid, but I'm perfectly fine with us making the decision to go with it until we discover otherwise.

If you want to look into moving assigned partitions into SubscriptionState, you're very welcome to. I agree that it's likely the way to go, but I don't think it necessarily has to be done within the scope of this PR.